Health and metrics exposure

robertkluin opened the pull request:
This adds the ability to enable telemetry / diagnostics endpoints within the controller. Eventually these may be used for health checks, but the immediately objective is to improve observability and help with debugging.

robertkluin committed:
Add queue for exposing telemetry from subsystems

The subsystems are (mostly) self-encapsulated and communicate via
queues. In order to expose internal system metrics and state a queue
will be used to export telemetry data. This will decouple the metrics
capture / exposure from the internal subsystems.

robertkluin committed:
Add optional health/telemetry capability

Adding telemetry in order to add health checks and aid in debugging.
This is disabled by default and the structures returned are alpha
quality so that we can better understand the practical usage of them.

robertkluin committed:
Return seconds-to-next-reconcile-attempt as formatted string

robertkluin committed:
Emit telemetry from the scheduler and event watcher

For the initial telemetry capture only the reconciliation scheduler and
event watcher subsystems emit data. This is rough and expected to evolve
after further testing and design.

robertkluin committed:
Bump Koreo Core version

robertkluin committed:
Add Starlette and uvicorn libs

robertkluin committed:
Update lockfile

ericlarssen commented on src/controller/__init__.py:
```diff
@@ -0,0 +1,219 @@
+            api=api,
+            namespace=namespace,
+            api_version=API_VERSION,
+            plural_kind=f"{kind_title.lower()}s",
```
Do we need to support the odd plural cases?

robertkluin replied:
Yes, normally. In this case these are only Koreo resources themselves (ResourceFunction, ResourceTemplate, ValueFunction, and Workflow)—all of which use simple plural rules. It is not currently an issue, but certainly could be if we add something with complex plural rules.

robertkluin committed:
Add jitter to scheduler's periodic checks

Without jitter, reconciliation of all resources a controller is
monitoring can become aligned. This is meant to help scatter that in
order to more effectively spread load.

robertkluin committed:
Ensure uvicorn starts before the controller

Starting uvicorn prior to the controller helps uvicorn start
successfully. The cause is not yet clear.

robertkluin committed:
Remove gratuitous try/excepts

Removing try/excepts in order to debug an issue with unclean uvicorn
shutdowns.

robertkluin committed:
Add logging for SystemExit

robertkluin committed:
Add more robust logging around core task groups

Python's TaskGroup has several special case errors. If these aren't
handled specially, then very verbose errors are dumped as the system
exits.

robertkluin committed:
Default to None, not an object

robertkluin committed:
Return on SystemExit or KeyboardInterrupt, raise CancelledError

robertkluin committed:
Make hot loading capable of toggling
